cpp/src/arrow/dataset/dataset_test.cc

```diff
@@ -50,28 +50,27 @@ TEST_F(TestInMemoryFragment, Scan) {
   AssertFragmentEquals(reader.get(), &fragment);
 }
 
-class TestInMemorySource : public DatasetFixtureMixin {};
+class TestInMemoryDataset : public DatasetFixtureMixin {};
 
-TEST_F(TestInMemorySource, GetFragments) {
+TEST_F(TestInMemoryDataset, GetFragments) {
   constexpr int64_t kBatchSize = 1024;
   constexpr int64_t kNumberBatches = 16;
 
   SetSchema({field("i32", int32()), field("f64", float64())});
   auto batch = ConstantArrayGenerator::Zeroes(kBatchSize, schema_);
   auto reader = ConstantArrayGenerator::Repeat(kNumberBatches, batch);
 
-  RecordBatchVector batches{static_cast<size_t>(kNumberBatches), batch};
-  auto fragment = std::make_shared<InMemoryFragment>(batches, options_);
   // It is safe to copy fragment multiple time since Scan() does not consume
   // the internal array.
-  auto source = InMemorySource(schema_, {static_cast<size_t>(kNumberBatches), batch});
+  auto dataset = std::make_shared<InMemoryDataset>(
```

Comment:
We'll need an `InMemoryDataset::Make` similarly like the rest of the datasets classes have to validate that the record batches have the same schema.

Reply:
`InMemoryDataset` is constructed with a generator; the vector constructor is a convenience overload (so the schema of contained record batches is not always available at construction time). Instead the schema is checked in `GetFragments`.

Reply:
Let's have a jira then where we can discuss further.

Reply:
I think we should have two factories for convenience:

```cpp
InMemoryDataset(Table)
InMemoryDataset(vector<RecordBatch>)  // all record batches should have the same schema
```

Reply:
https://issues.apache.org/jira/browse/ARROW-7945?focusedCommentId=17045585&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17045585